src/RollerPeriphery.sol

```diff
@@ -43,44 +76,55 @@ contract RollerPeriphery is Trust {
+    error InvalidExchangeProxy();
+
+    constructor(IPermit2 _permit2, address _exchangeProxy) Trust(msg.sender) {
+        permit2 = _permit2;
+        exchangeProxy = _exchangeProxy;
     }
 
-    /// @notice Redeem vault shares and convert to underlying, with slippage protection
+    /// @notice Redeem vault shares to any token, with slippage protection
     /// @param roller AutoRoller vault
     /// @param shares Number of shares to redeem
     /// @param receiver Destination address for the returned assets
-    /// @param minAmountOut Minimum amount of underlying returned
-    /// @return underlyingOut Amount of underlying converted from Target redeemable by the given number of shares
-    function redeemForUnderlying(AutoRoller roller, uint256 shares, address receiver, uint256 minAmountOut) external returns (uint256 underlyingOut) {
-        AdapterLike adapter = AdapterLike(address(roller.adapter()));
-
-        if ((underlyingOut = adapter.unwrapTarget(roller.redeem(shares, address(this), msg.sender))) < minAmountOut) {
-            revert MinUnderlyingError();
+    /// @param minAmountOut Minimum amount of tokens returned // TODO: check
+    /// @param permit Permit message to pull shares from caller
+    /// @param quote Swap quote for converting underlying to token
+    /// @return amtOut Amount of tokens redeemed by the given number of shares
+    function redeem(AutoRoller roller, uint256 shares, address receiver, uint256 minAmountOut, PermitData calldata permit, SwapQuote calldata quote) external returns (uint256 amtOut) {
+        _transferFrom(permit, address(roller), shares);
+
+        if ((amtOut = _fromTarget(address(roller.adapter()), roller.redeem(shares, address(this), address(this)), quote)) < minAmountOut) {
+            revert MinAmountOutError();
         }
+        address(quote.buyToken) == ETH
+            ? payable(receiver).transfer(amtOut)
+            : ERC20(address(quote.buyToken)).safeTransfer(receiver, amtOut); // transfer bought tokens to receiver
 
-        ERC20(adapter.underlying()).safeTransfer(receiver, underlyingOut);
+        _transferUnderlying(roller, receiver);
     }
 
     /// @notice Withdraw asset from vault with slippage protection
     /// @param roller AutoRoller vault
     /// @param assets Amount of asset requested for withdrawal
     /// @param receiver Destination address for the returned assets
     /// @param maxSharesOut Maximum amount of shares burned
+    /// @param permit Permit message to pull shares from caller
     /// @return shares Number of shares to redeem
-    function withdraw(AutoRoller roller, uint256 assets, address receiver, uint256 maxSharesOut) external returns (uint256 shares) {
-        if ((shares = roller.withdraw(assets, receiver, msg.sender)) > maxSharesOut) {
+    function withdrawTarget(AutoRoller roller, uint256 assets, address receiver, uint256 maxSharesOut, PermitData calldata permit) external returns (uint256 shares) {
```

Comment:
we've decided not to use zaps neither `withdraw` nor `minting` because it was a bit tricky to make it work. 
E.g let's say you have `maDAI` as target, `DAI` as underlying and you want to withdraw to `USDC`.

The call to `withdraw()` would need to receive the exact amount of `USDC` you want to withdraw and then the flow would be something like:
1. Calculate how much is the amount of `USDC` in target (`maDAI`) (off-chain and sent via param) 
2. Pull the amount of `USDC`
3. Zap `USDC` to underlying (`DAI`) and wrap to convert it to target (`maDAI`)
4. Call `roller.withdraw()` with the calculated amount in 1, which will pull the `maDAI`. The amount that this function tries to pull needs to be the same (or a bit less) than what the Periphery is currently holding. And this is tricky since we can't guarantee that amount swapped in 3. 
Because of this and some other nuances we've decided it was better not to allow zaps on these functions.

Reply:
Hmm! I'm confused because I thought the roller holds maDAI/YTs/PTs.

I thought it would be something like:
1. preview the amount of DAI received from `withdraw` (using `previewWithdraw`)
2. Generate quote for DAI -> USDC
3. Call `roller.withdraw`, which withdraws `maDAI`, unwraps to `DAI`, and calls the zap with the received quote?